src/ui/FacetValueSuggestions/FacetValueSuggestions.ts

```diff
@@ -0,0 +1,350 @@
+          keyword: fieldResponse.keyword,
+          value: indexFieldValue.value,
+          score: this.computeScoreForSuggestionRow(indexFieldValue, fieldTotalReference)
+        };
+      });
+      return allValues.concat(suggestionRows);
+    }, []);
+
+    this.logger.debug('FacetValue Suggestions Results', resultsWithoutSelectedValues);
+
+    const rankedResults = resultsWithoutSelectedValues.sort(
+      (a, b) => b.score.distanceFromTotalForField - a.score.distanceFromTotalForField
+    );
+    const preciseResults = rankedResults.splice(0, Math.ceil(this.options.numberOfSuggestions / 2));
+    const broadResults = rankedResults.slice(-1, Math.floor(this.options.numberOfSuggestions / 2));
+
+    return [].concat(preciseResults).concat(broadResults);
+  }
+
+  private getQuerySuggestionsKeywords(omnibox: Omnibox): Promise<string[]> {
+    if (this.options.useQuerySuggestions && omnibox.suggestionAddon) {
+      const suggestions = omnibox.suggestionAddon.getSuggestion();
+      return suggestions ? suggestions.then(suggestions => suggestions.map(s => s.text)) : Promise.resolve();
+    } else {
+      return Promise.resolve();
+    }
+  }
+
+  private computeScoreForSuggestionRow(fieldValue: IIndexFieldValue, reference: IFacetValueReference): IFacetValueSuggestionScore {
+    const totalNumberForFieldValue = reference.fieldsTotal[fieldValue.value] || reference.smallestTotal;
+    const distanceFromTotalForField: number = (totalNumberForFieldValue - fieldValue.numberOfResults) / totalNumberForFieldValue * 100;
+    return {
+      distanceFromTotalForField: distanceFromTotalForField
+    };
+  }
+
+  private listFieldValuesBatch(valuesToSearch: string[]): Promise<IFacetValueSuggestionsResponse> {
+    // The reference request will be used to get the maximum number of values for a given facet value.
+    const referenceValuesRequest = this.buildReferenceFieldValueRequest();
+    const queryParts = this.getQueryToExecuteParts();
+    const suggestionValuesRequests = valuesToSearch.map(value => this.buildListFieldValueRequest(queryParts.concat(value).join(' ')));
+    const requests = [].concat(suggestionValuesRequests).concat(referenceValuesRequest);
+    return this.queryController
+      .getEndpoint()
+      .listFieldValuesBatch({
+        batch: requests
+      })
+      .then(values => {
+        const reference = this.computeReferenceFromBatch(values.pop());
+        const remainingResponses: IFacetValueBatchResponse[] = values.map((value, i) => {
+          return <IFacetValueBatchResponse>{
+            keyword: valuesToSearch[i],
+            values: value
+          };
+        });
+        return <IFacetValueSuggestionsResponse>{
+          responses: remainingResponses,
+          reference: reference
+        };
+      });
+  }
+
+  private computeReferenceFromBatch(batch: IIndexFieldValue[]): IFacetValueReference {
+    const reference: IFacetValueReference = {
+      fieldsTotal: {},
```

Comment:
Map the `batch` parameter to the needed `reference` structure, instead of building it and then doing a forEach loop.